### PR TITLE
Add model wrapper for wrapping data and predictions

### PR DESCRIPTION
**Pull request by gaugup**
The model wrapper base class `ModelWrapperPredictions` wraps the test data and model prediction and supports the predict call. The `ModelWrapperPredictionsClassification` also supports the `predict_proba` method.

Signed-off-by: Gaurav Gupta <gaugup@microsoft.com>

### CODE REVIEW

**Review by vinuthakaranth — APPROVED** [review]

File: python/ml_wrappers/model/predictions_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                "Expecting a pandas dataframe for query_test_data")
+        prediction_output = []
+        for _, row in query_test_data.iterrows():
+            filtered_df = self._get_filtered_data(row)
+            prediction_output.append(filtered_df[TARGET].values[0])
+
+        return np.array(prediction_output)
+
+    def __setstate__(self, state):
+        """Set the state so that the wrapped model is
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/predictions_wrapper.py:88]: do you want to add the state param in the doc here

File: python/ml_wrappers/model/predictions_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        serializable via pickle."""
+        self._combined_data = state["_combined_data"]
+
+    def __getstate__(self):
+        """Return the state so that the wrapped model is
+        serializable via pickle."""
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/predictions_wrapper.py:94]: do you want to add the return value/type in the doc here

File: python/ml_wrappers/model/predictions_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+
+TARGET = 'target'
+
+
+class ModelWrapperPredictions:
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/predictions_wrapper.py:15]: nit: maybe PredictionsModelWrapper is a better name for the class than ModelWrapperPredictions?

File: python/ml_wrappers/model/predictions_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    """Model wrapper to wrap the samples used to train the models
+    and the predictions of the model. This wrapper is useful when
+    it is not possible to load the model.
+    """
+    def __init__(self, test_data: pd.DataFrame, y_pred: np.ndarray):
+        """Creates an ModelWrapper object.
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/predictions_wrapper.py:21]: nit: Creates a ModelWrapperPredictions object.

File: python/ml_wrappers/model/predictions_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    def predict(self, query_test_data: pd.DataFrame) -> np.ndarray:
+        """Return the predictions based on the query data.
+
+        :param query_test_data: The data for which the predictions need to
+            be returned.
+        :type query_test_data: pd.DataFrame
+        :return: Predictions of the model.
+        :rtype: np.ndarray
+        """
+        if not isinstance(query_test_data, pd.DataFrame):
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/predictions_wrapper.py:77]: Looks great.  Did you want to use KNN instead, with nn=1, or keep the filtering logic here?

**Review by imatiach-msft — APPROVED** [review]

File: python/ml_wrappers/model/predictions_wrapper.py (hunk 3 above)
Comment (gaugup) [reply python/ml_wrappers/model/predictions_wrapper.py:15]: Done. Thanks for the suggestion.

File: python/ml_wrappers/model/predictions_wrapper.py (hunk 4 above)
Comment (gaugup) [reply python/ml_wrappers/model/predictions_wrapper.py:21]: Done

File: python/ml_wrappers/model/predictions_wrapper.py (hunk 5 above)
Comment (gaugup) [reply python/ml_wrappers/model/predictions_wrapper.py:77]: The Knn could be used to replace the current search based approach. But there are few things to consider:-

1. We need to transform the string features as knn cannot take string features https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
2. We need to think how we want to model the problem as a classification problem with as many classes as number of instances of test data. We need to do more testing if knn could work with such high number of classes. 

Maybe we can merge this and do a separate investigation on using knn inside of the search.

File: python/ml_wrappers/model/predictions_wrapper.py (hunk 1 above)
Comment (gaugup) [reply python/ml_wrappers/model/predictions_wrapper.py]: Done

File: python/ml_wrappers/model/predictions_wrapper.py (hunk 2 above)
Comment (gaugup) [reply python/ml_wrappers/model/predictions_wrapper.py]: Done

**Pull request merged**